design/object-store-realm.md

```diff
@@ -0,0 +1,214 @@
+
+This example `object-store-realm.yaml`, names a realm with 2 zonegroups.
+```yaml
+apiVersion: ceph.rook.io/v1alpha1
+kind: CephObjectStoreRealm
+metadata:
+  name: my-realm
+  namespace: rook-ceph
+spec:
+  bootstrap-realm:
+    name: my-realm
+    zonegroups:
+      my-zonegroup-a:
+        is-master: true
+        master-zone: zone-z
+      my-zonegroup-b:
+        is-master: false
+        master-zone: zone-y
+```
+
+Now create the object store realm.
+```bash
+kubectl create -f object-store-realm.yaml
+```
+
+#### Steps
+
+At this point the Rook operator recognizes that a new object store realm resource needs to be configured. The operator will create all of the resources to start the object store realm.
+
+1. A Ceph realm is created if it does not exist already
+2. A Ceph master zonegroup is created in the realm if it does not exist already
+3. Any non-master Ceph zonegroup is created in the realm if it does not exist already
+4. A master zone is created for the zonegroup.
+
+After these steps the admin should start up object stores configured with the same zone, zonegroup, and realm as configured (see next section).
+
+#### Notes
+1. The zonegroups section of the object-store-realm CR is only additive. If the playbook is reapplied and zonegroup already exist the zonegroup won't be recreated. If a zonegroup previously created is removed from the list of zonegroups, it will not be deleted. If a new zonegroup is added to the list it will be created. 
+2. The same principle applies to master-zones.
+3. When resource is deleted, realm and zonegroups are not deleted from the cluster. Realm and zonegroup deletion must be done through toolboxes.
+
+### Creating a Zone on the same cluster the object-store-realm was created.
+
+Once the storage admin creates a realm and zonegroup(s) on a cluster, the admin is ready to add object-stores to a master zone on a realm or add another zone to the zonegroup. 
+The admin will specify his/her configuration settings in the realm section of the object-store like in the following `object-store.yaml`.
+
+All objectstores with the realm section must be named in the form `realm_name` + `zone_name`.`zone_name` can be the name of a master zone created in the object store realm CR or the name of any other one..
+
+If the name of the objectstore is `realm_name` + '-' + `master_zone_name`, the the object stores created will be added to the master zone.
+
+If the name of the objectstore is not `realm_name` + '-' + `master_zone_name`, then a new zone will be created with the object stores created via the resource.
+
+This example config will add objectstores to the already created zone `zone-y`.
+These objectstores will go into the master zone, zonegroup, and realm created for the ObjectStoreRealm resource above.
+```yaml
+apiVersion: ceph.rook.io/v1alpha1
+kind: CephObjectStore
+metadata:
+  name: my-realm-zone-y
```

Comment:
This point starts with "Once the storage admin creates a realm and zonegroup(s) on a cluster, the admin is ready to add object-stores to a master zone on a realm or add another zone to the zonegroup."
Now you are using the "name" attribute of the realm of the CRD for the name of the zone. i mean now this CRD attribute has two different uses...

Besides that the different elements of the RGW topology are not reflected properly in k8's, and this also make difficult to understand how specifications, and how changes can affect.
I would suggest to have the following k8s objects:


**For example, the object store will be:**

```
kind: CephObjectStore
metadata:
  name: my-store
  namespace: rook-ceph
spec:
  metadataPool:
    failureDomain: host
    replicated:
      size: 3
  dataPool:
    failureDomain: device
    erasureCoded:
      dataChunks: 6
      codingChunks: 2
  gateway:
    type: s3
    port: 80
    securePort: 443
    instances: 3
```

**The realm:**

```
kind: CephObjectStoreRealm
metadata:
  name: my-realm
  namespace: rook-ceph
spec:
  bootstrap-realm: ?¿?¿ (i do not know what is this attribute)
  name: my-realm
```

**The zone group:**

```
kind: CephObjectStoreZoneGroup
metadata:
    name: my-zone-group 
spec:
    realm: my-realm
    is-master: true
```

**The zone:**

```
kind: CephObjectStoreZone
metadata:
    name: my-zone
spec:
    zone-group: my-zone-group 
    is-master: true
    pull-endpoint: http://rook-ceph-my-realm-zone-y:80
```

Once we have this basic set of object, i think that it will be more easy to create CRD files for different uses:

Example:
- create the object store with a default configuration ( 1 realm, 1 master zone group, 1 master zone):

```
kind: CephObjectStore
...
kind: CephObjectStoreRealm
...
kind: CephObjectStoreZoneGroup
...
kind: CephObjectStoreZone
```

- create a new zone in a existing zone group:

```
kind: CephObjectStoreZone
metadata:
    name: my-new-zone
spec:
    zone-group: my-zone-group 
    is-master: false
    pull-endpoint:
```

- create a new zone in a new zone group synced with the remote zone-y:
```
kind: CephObjectStoreZoneGroup
metadata:
    name: my-new-zone-group 
spec:
    realm: my-realm
    is-master: false
kind: CephObjectStoreZone
metadata:
    name: my-new-zone
spec:
    zone-group: my-new-zone-group  
    is-master: false
    pull-endpoint: http://external-my-realm-zone-y:80
```

Reply:
@jmolmo This was my initial design. After many working sessions with Sage, Casey, Travis, and I this approach fell out of favor for many reasons.
Some include:
- Having a different resource for every multisite element would lead to a bunch of resources getting created. In our design there's only 1 per realm.
-If you have this CR set up on one Ceph cluster and want to sync object-store from another, all you have to do is add the `realm` section in the object-strore config with `pull-endpoint` defined. No additional resources need to be created.
- The initial period cannot be committed unless a realm, zonegroup, and master zone have already been created. Everytime there's a change made related to some state in the realm the period needs to be committed.  
- In this design, the object-store acts somewhat like the zone CR. If `realm` is specified in the object-store config than a zone is created with the same name as the object-store. If the name is the same as the master zone in the Object-Store-Realm CR then the endpoints for those object-stores are added to the master zone. 

I'd be happy to chat and go into more details.

Reply:
> @jmolmo This was my initial design. After many working sessions with Sage, Casey, Travis, and I this approach fell out of favor for many reasons.
> Some include:
> 
> * Having a different resource for every multisite element would lead to a bunch of resources getting created. In our design there's only 1 per realm.
usually, in k8s, having different resource for every element is a good thing

>   -If you have this CR set up on one Ceph cluster and want to sync object-store from another, all you have to do is add the `realm` section in the object-strore config with `pull-endpoint` defined. No additional resources need to be created.
> * The initial period cannot be committed unless a realm, zonegroup, and master zone have already been created. Everytime there's a change made related to some state in the realm the period needs to be committed.

there is a difference between committing stuff and defining stuff. creation of a CRD does not mean that it takes effect immediately, it just mean that the configuration is there, and the software try to get the current state to the desired one

> * In this design, the object-store acts somewhat like the zone CR. If `realm` is specified in the object-store config than a zone is created with the same name as the object-store. If the name is the same as the master zone in the Object-Store-Realm CR then the endpoints for those object-stores are added to the master zone.
> 
> I'd be happy to chat and go into more details.